ledger-core/v2/virtual/request/vstaterequest.go

```diff
@@ -0,0 +1,173 @@
+	"github.com/insolar/assured-ledger/ledger-core/v2/insolar/payload"
+	"github.com/insolar/assured-ledger/ledger-core/v2/network/messagesender"
+	messageSenderAdapter "github.com/insolar/assured-ledger/ledger-core/v2/network/messagesender/adapter"
+	"github.com/insolar/assured-ledger/ledger-core/v2/vanilla/injector"
+	"github.com/insolar/assured-ledger/ledger-core/v2/vanilla/throw"
+	"github.com/insolar/assured-ledger/ledger-core/v2/virtual/object"
+)
+
+type SMVStateRequest struct {
+	// input arguments
+	Meta    *payload.Meta
+	Payload *payload.VStateRequest
+
+	objectStateReport *payload.VStateReport
+
+	failReason payload.VStateUnavailable_ReasonType
+
+	messageSender *messageSenderAdapter.MessageSender
+	pulseSlot     *conveyor.PulseSlot
+}
+
+/* -------- Declaration ------------- */
+
+var dSMVStateRequestInstance smachine.StateMachineDeclaration = &dSMVStateRequest{}
+
+type dSMVStateRequest struct {
+	smachine.StateMachineDeclTemplate
+}
+
+func (*dSMVStateRequest) InjectDependencies(sm smachine.StateMachine, _ smachine.SlotLink, injector *injector.DependencyInjector) {
+	s := sm.(*SMVStateRequest)
+
+	injector.MustInject(&s.pulseSlot)
+	injector.MustInject(&s.messageSender)
+}
+
+func (*dSMVStateRequest) GetInitStateFor(sm smachine.StateMachine) smachine.InitFunc {
+	s := sm.(*SMVStateRequest)
+	return s.Init
+}
+
+/* -------- Instance ------------- */
+
+func (*SMVStateRequest) InjectDependencies(sm smachine.StateMachine, _ smachine.SlotLink, injector *injector.DependencyInjector) {
+	s := sm.(*SMVStateRequest)
+
+	injector.MustInject(&s.pulseSlot)
+	injector.MustInject(&s.messageSender)
+}
+
+func (s *SMVStateRequest) GetStateMachineDeclaration() smachine.StateMachineDeclaration {
+	return dSMVStateRequestInstance
+}
+
+func (s *SMVStateRequest) Init(ctx smachine.InitializationContext) smachine.StateUpdate {
+	return ctx.Jump(s.stepProcess)
+}
+
+func (s *SMVStateRequest) stepProcess(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	var (
+		objectCatalog = object.Catalog{}
+	)
+	objectSharedState, stateFound := objectCatalog.TryGet(ctx, s.Payload.Callee)
+	if !stateFound {
+		s.failReason = payload.Missing
+		return ctx.Jump(s.stepReturnStateUnavailable)
+	}
+
+	var (
+		semaphoreReadyToWork smachine.SyncLink
+	)
+
+	action := func(state *object.SharedState) {
+		semaphoreReadyToWork = state.ReadyToWork
+
+		descriptor := state.Descriptor()
+
+		if descriptor == nil {
+			panic(throw.IllegalState())
+		}
+
+		stateRef := descriptor.HeadRef()
+
+		immutableCounts, _ := state.ImmutableExecute.GetCounts()
+		mutableCounts, _ := state.MutableExecute.GetCounts()
+		s.objectStateReport = &payload.VStateReport{
+			AsOf:                  s.Payload.AsOf,
+			Callee:                s.Payload.Callee,
+			ImmutablePendingCount: int32(immutableCounts),
+			MutablePendingCount:   int32(mutableCounts),
+			LatestDirtyState:      *stateRef,
+		}
+
+		if s.Payload.RequestedContent^payload.RequestLatestDirtyState == 0 {
```

Comment:
Might be is it better to add function Contains() to StateRequestContentFlags

Reply:
done